pkg/admission/crdnooverlappinggvr/crdnooverlappinggvr_admission.go

```diff
@@ -0,0 +1,106 @@
+		func(_ io.Reader) (admission.Interface, error) {
+			return &crdNoOverlappingGVRAdmission{
+				Handler: admission.NewHandler(admission.Create),
+			}, nil
+		})
+}
+
+type crdNoOverlappingGVRAdmission struct {
+	*admission.Handler
+
+	apiBindingLister apislisters.APIBindingLister
+}
+
+// Ensure that the required admission interfaces are implemented.
+var _ = admission.ValidationInterface(&crdNoOverlappingGVRAdmission{})
+
+func (p *crdNoOverlappingGVRAdmission) SetKcpInformers(informers kcpinformers.SharedInformerFactory) {
+	p.SetReadyFunc(informers.Apis().V1alpha1().APIBindings().Informer().HasSynced)
+	p.apiBindingLister = informers.Apis().V1alpha1().APIBindings().Lister()
+}
+
+// Validate checks if the given CRD's Group and Resource don't overlap with bound CRDs
+func (p *crdNoOverlappingGVRAdmission) Validate(ctx context.Context, a admission.Attributes, _ admission.ObjectInterfaces) error {
+	if a.GetResource().GroupResource() != apiextensions.Resource("customresourcedefinitions") {
+		return nil
+	}
+	if a.GetKind().GroupKind() != apiextensions.Kind("CustomResourceDefinition") {
+		return nil
+	}
+	clusterName, err := request.ClusterNameFrom(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve cluster from context: %w", err)
+	}
+	// ignore CRDs targeting system workspaces
+	if clusterName == apibinding.ShadowWorkspaceName || clusterName == logicalcluster.New("system:admin") {
+		return nil
+	}
+
+	crd, ok := a.GetObject().(*apiextensions.CustomResourceDefinition)
+	if !ok {
+		return fmt.Errorf("unexpected type %T", a.GetObject())
+	}
+	apiBindingsForCurrentClusterName, err := p.listAPIBindingsFor(clusterName)
+	if err != nil {
+		return err
+	}
+	var notReadyAPIBindings []string
+	for _, apiBindingForCurrentClusterName := range apiBindingsForCurrentClusterName {
+		if !conditions.IsTrue(apiBindingForCurrentClusterName, apisv1alpha1.BindingUpToDate) {
+			notReadyAPIBindings = append(notReadyAPIBindings, apiBindingForCurrentClusterName.Name)
+			// give it a try anyway, we fail on not ready bindings
+		}
+		for _, boundResource := range apiBindingForCurrentClusterName.Status.BoundResources {
+			if boundResource.Group == crd.Spec.Group && boundResource.Resource == crd.Spec.Names.Plural {
+				return admission.NewForbidden(a, fmt.Errorf("cannot create %q crd with %q group and %q resource because it overlaps with a bound crd for %q apiBinding in %q logical cluster",
+					crd.Name, crd.Spec.Group, crd.Spec.Names.Plural, apiBindingForCurrentClusterName.Name, clusterName))
+			}
+		}
+	}
+	if len(notReadyAPIBindings) > 0 {
+		return admission.NewForbidden(a, fmt.Errorf("cannot create %q crd because the following apiBindings werern't ready and werern't checked for conflicting GVRs = %v", crd.Name, notReadyAPIBindings))
+	}
+	return nil
+}
+
+func (p *crdNoOverlappingGVRAdmission) listAPIBindingsFor(clusterName logicalcluster.Name) ([]*apisv1alpha1.APIBinding, error) {
+	list, err := p.apiBindingLister.List(labels.Everything())
```

Comment:
use an index as in https://github.com/kcp-dev/kcp/blob/main/pkg/reconciler/scheduling/placement/placement_indexes.go#L30.